File: backend/src/infra/verification/sandbox.ts
import { execFile, exec } from "node:child_process";
import { promisify } from "node:util";
import { readFile } from "node:fs/promises";
import path from "node:path";
import { getErrorMessage } from "../../utils/error.js";
import { validatePathBoundary } from "../security/path-guard.js";

const execFileAsync = promisify(execFile)
const execAsync = promisify(exec)

const TIMEOUT_MS = 5 * 60 * 1000
const MAX_OUTPUT_BYTES = 10 * 1024 * 1024
const SANDBOX_IMAGE = "node:20-slim"
const PREPARE_PNPM = "corepack enable && corepack prepare pnpm@latest --activate"

export interface CommandResult {
    passed: boolean
    output: string
}

export async function installDependencies(repoPath: string): Promise<CommandResult> {
    return runInSandbox(repoPath, `${PREPARE_PNPM} && pnpm install`, { allowNetwork: true })
}

/**
 * Menjalankan `npx tsc --noEmit`.
 * Jika `isLocalMode === true`, dieksekusi langsung di host PC (`cwd: repoPath`).
 * Jika `isLocalMode === false`, dieksekusi di dalam container Docker sandbox.
 */
export async function runTypeCheck(repoPath: string, isLocalMode = false): Promise<CommandResult> {
    const validatedPath = validatePathBoundary(repoPath, repoPath)
    if (isLocalMode) {
        return runLocalHostCommand(validatedPath, "npx tsc --noEmit")
    }
    return runInSandbox(validatedPath, "npx tsc --noEmit", { allowNetwork: false })
}

/**
 * Menjalankan skrip "lint" dan "test" milik repositori.
 * Jika `isLocalMode === true`, dieksekusi langsung di host PC (`cwd: repoPath`).
 * Jika `isLocalMode === false`, dieksekusi di dalam container Docker sandbox.
 */
export async function runLinterAndTests(repoPath: string, isLocalMode = false): Promise<CommandResult> {
    const validatedPath = validatePathBoundary(repoPath, repoPath)
    const scripts = await getPackageScripts(validatedPath)
    const commands: string[] = []

    if (scripts.lint) commands.push("pnpm run lint")
    if (scripts.test) commands.push("pnpm test")

    if (commands.length === 0) {
        return { passed: true, output: "Tidak ditemukan skrip lint atau test di package.json - dilewati." }
    }

    const commandStr = commands.join(" && ")

    if (isLocalMode) {
        return runLocalHostCommand(validatedPath, commandStr)
    }

    return runInSandbox(validatedPath, `${PREPARE_PNPM} && ${commandStr}`, { allowNetwork: false })
}

async function getPackageScripts(repoPath: string): Promise<Record<string, string>> {
    try {
        const raw = await readFile(path.join(repoPath, "package.json"), "utf-8")
        return JSON.parse(raw).scripts ?? {}
    } catch {
        return {}
    }
}

async function runLocalHostCommand(repoPath: string, command: string): Promise<CommandResult> {
    try {
        const { stdout, stderr } = await execAsync(command, {
            cwd: repoPath,
            timeout: TIMEOUT_MS,
            maxBuffer: MAX_OUTPUT_BYTES,
        })
        return { passed: true, output: stdout + stderr }
    } catch (error) {
        return { passed: false, output: getErrorMessage(error) }
    }
}

async function runInSandbox(
    repoPath: string,
    shellCommand: string,
    { allowNetwork }: { allowNetwork: boolean }
): Promise<CommandResult> {
    const args = [
        "run", "--rm",
        ...(allowNetwork ? [] : ["--network", "none"]),
        "--memory", "2g",
        "--cpus", "2",
        "--pids-limit", "256",
        "-v", `${repoPath}:/repo`,
        "-w", "/repo",
        SANDBOX_IMAGE,
        "sh", "-c", shellCommand,
    ]

    try {
        const { stdout, stderr } = await execFileAsync("docker", args, {
            timeout: TIMEOUT_MS,
            maxBuffer: MAX_OUTPUT_BYTES,
        })
        return { passed: true, output: stdout + stderr }
    } catch (error) {
        return { passed: false, output: getErrorMessage(error) }
    }
}

File: engine/app/infra/verification/sandbox.py
"""1:1 port dari runTypeCheck/runLinterAndTests di backend/src/infra/verification/sandbox.ts.

installDependencies() SENGAJA gak diporting kesini - itu tetap permanen di TypeScript
(dipicu admin pas sync repo, bukan tool yang dipanggil agent)."""

import asyncio
import json
import logging
from pathlib import Path

from app.infra.proc import CommandResult, run_subprocess
from app.infra.security.path_guard import validate_path_boundary

logger = logging.getLogger(__name__)

TIMEOUT_SECONDS = 5 * 60
MAX_OUTPUT_BYTES = 10 * 1024 * 1024


async def run_type_check(repo_path: str) -> CommandResult:
    """Jalanin `npx tsc --noEmit` di host (cwd=repo_path). Gak ada Docker lagi - repo selalu lokal."""
    logger.info("running type check: repo_path=%s", repo_path)
    validated = validate_path_boundary(repo_path, repo_path)
    result = await run_subprocess(
        ["npx tsc --noEmit"], cwd=validated, timeout_s=TIMEOUT_SECONDS,
        max_output_bytes=MAX_OUTPUT_BYTES, shell=True,
    )
    logger.info("type check completed: passed=%s output_length=%s", result.passed, len(result.output))
    return result


async def run_linter_and_tests(repo_path: str) -> CommandResult:
    """Jalanin skrip lint & test (pnpm) di host (cwd=repo_path). Gak ada Docker lagi."""
    logger.info("running linter/tests: repo_path=%s", repo_path)
    validated = validate_path_boundary(repo_path, repo_path)
    scripts = await _get_package_scripts(validated)
    commands = [
        cmd for cmd, present in (
            ("pnpm run lint", scripts.get("lint")),
            ("pnpm test", scripts.get("test")),
        ) if present
    ]

    if not commands:
        logger.info("no lint/test scripts found in package.json - skipped")
        return CommandResult(True, "Tidak ditemukan skrip lint atau test di package.json - dilewati.")

    command_str = " && ".join(commands)
    logger.info("executing commands: %s", command_str)
    result = await run_subprocess(
        [command_str], cwd=validated, timeout_s=TIMEOUT_SECONDS,
        max_output_bytes=MAX_OUTPUT_BYTES, shell=True,
    )
    logger.info("linter/tests completed: passed=%s output_length=%s", result.passed, len(result.output))
    return result


async def _get_package_scripts(repo_path: str) -> dict:
    try:
        raw = await asyncio.to_thread((Path(repo_path) / "package.json").read_text, encoding="utf-8")
        return json.loads(raw).get("scripts") or {}
    except Exception:
        return {}
